R/fixest-tidiers.R

```diff
@@ -0,0 +1,197 @@
+#' @templateVar class fixest
+#' @template title_desc_tidy
+#'
+#' @param x A `fixest` object returned from any of the `fixest` estimators
+#' @template param_confint
+#' @param ... Additional arguments passed to `summary` and `confint`. Important
+#' arguments are `se` and `cluster`. See [`summary.fixest`][fixest::summary.fixest()].
+#' @evalRd return_tidy(regression = TRUE)
+#'
+#' @details The `fixest` package provides a family of functions for estimating
+#'   models with arbitrary numbers of fixed-effects, in both an OLS and a GLM
+#'   context. The package also supports robust (i.e. White) and clustered
+#'   standard error reporting via the generic `summary.fixest()` command. In a
+#'   similar vein, the `tidy()` method for these models allows users to specify
+#'   a desired standard error correction either 1) implicitly via the supplied
+#'   fixest object, or 2) explicitly as part of the tidy call. See examples
+#'   below.
+#'
+#' @examples
+#' \donttest{
+#' library(fixest)
+#' gravity <- feols(log(Euros) ~ log(dist_km) | Origin + Destination + Product + Year, trade)
+#'
+#' tidy(gravity)
+#' glance(gravity)
+#' augment(gravity, trade)
+#'
+#' ## To get robust or clustered SEs, users can either:
+#' # 1) Feed tidy() a summary.fixest object that has already accepted these arguments
+#' gravity_summ <- summary(gravity, cluster = c("Product", "Year"))
+#' tidy(gravity_summ, conf.int = TRUE)
+#' # 2) Or, specify the arguments directly in the tidy() call
+#' tidy(gravity, conf.int = TRUE, cluster = c("Product", "Year"))
+#' tidy(gravity, conf.int = TRUE, se = "threeway")
+#' # etc.
+#'
+#' ## The other fixest methods all working similarly. For example:
+#' gravity_pois <- feglm(Euros ~ log(dist_km) | Origin + Destination + Product + Year, trade)
+#' tidy(gravity_pois)
+#' glance(gravity_pois)
+#' augment(gravity_pois, trade)
+#' }
+#'
+#' @export
+#' @family fixest tidiers
+#' @seealso [tidy()], [fixest::feglm()], [fixest::fenegbin()],
+#' [fixest::feNmlm()], [fixest::femlm()], [fixest::feols()], [fixest::fepois()]
+tidy.fixest <- function(x, conf.int = FALSE, conf.level = 0.95, ...) {
+  has_dots <- !missing(...)
+  # These two options are necessary because the user might specify
+  # broom(x, cluster="something") or
+  # broom(summary(x, cluster="something"))
+  # and calling summary() again will overwrite the arguments in the second case.
+  # Note that class(summary(x)) == class(x) == "fixest".
```

Comment:
Big oooof. This is rough to deal with. Okay, so are you checking if the input object is `summary(fixest)` or `fixest` based on the other arguments the user gives this function? That seems brittle. I would prefer logic along the lines of

```
tidy.fixest {
  # figure out if we have have a `fixest` or a `summary(fixest`) object
  # call an appropriate helper
}

tidy_summary_fixest <- function(...) {
  # do stuff
}

tidy_fixest <- function(...) {
  # do other stuff
}
```

and then some error if the passed arguments and the passed `fixest` are incompatible.